### PR TITLE
Add BWC for batch ingestion

### PR DESCRIPTION
**Pull request by chishui**
### Description
Add BWC tests for batch ingestion feature.

### Issues Resolved
https://github.com/opensearch-project/neural-search/issues/763

### Check List
- [x] New functionality includes testing.
    - [x] All tests pass
- [x] New functionality has been documented.
    - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: qa/restart-upgrade/build.gradle
Comment (yuye-aws) [inline qa/restart-upgrade/build.gradle:93]: this feature -> these features

File: qa/restart-upgrade/build.gradle
Comment (yuye-aws) [inline qa/restart-upgrade/build.gradle:150]: this feature -> these features

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1298]: Why are we asserting that ingestOutputs is an instance of Map? I think dense embedding should be a vector, which is a list of numbers.

File: src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java
Comment (yuye-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java:14]: help class -> helper class

File: qa/rolling-upgrade/build.gradle
Comment (yuye-aws) [inline qa/rolling-upgrade/build.gradle:93]: this feature -> these features

File: qa/rolling-upgrade/build.gradle
Comment (yuye-aws) [inline qa/rolling-upgrade/build.gradle:151]: this feature -> these features

File: qa/rolling-upgrade/build.gradle
Comment (yuye-aws) [inline qa/rolling-upgrade/build.gradle:209]: Do you know why `k-NN radial search` is not excluded here?

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:694]: Do we really need the `id` field from the docs? If not, how about making `docs` a list of strings? 
From this function, we implicitly assert that every element in the list contains two keys, i.e., `id` and `text`. Other users might be confused with the type of `docs`

File: src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java
Comment (yuye-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java:30]: If function `addDocsThroughBulk` does not need the `id` field, do we still need this function?

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1299]: Thanks for creating this function in the base class. How about making this function more general? The function will only check whether a certain field exists for every document in the index. If so, it will return a list with elements in this field. The returned list can be used for further evaluation.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1291]: How about using the `getFirstInnerHit` function? Besides, do we need to validate every document from the index?

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1298]: We use sparse encoding to test batch ingestion so verify map type here.

File: qa/rolling-upgrade/build.gradle
Comment (chishui) [reply qa/rolling-upgrade/build.gradle:209]: That's a good question, @junqiu-lei do you happen to know why  `k-NN radial search` is not excluded under some rolling upgrade tasks? Is it intentional?

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:694]: We also want to verify if a certain doc exists after bulk action through querying doc by known id.

File: src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java:30]: answered above

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1299]: That's a good callout. But I feel it makes a bit more sense that we modify it to the recommended manner when we do have the need, it won't be a fully restructure but a couple of lines change.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1291]: Single doc retrieval doesn't have the `hits` field in response. We verify the total count first that means we successfully ingest all the documents, then we pick up one doc to verify it's detailed information, doesn't seem necessary to verify them all.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1298]: This function is from the base class, so we cannot assume this function will only be used for batch ingestion.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1299]: In fact, this function can be used by text chunking BWC tests files. Can you take a look at the `validateTestIndex` function? These two functions are pretty similar and we can reuse them.

File: qa/rolling-upgrade/build.gradle
Comment (yuye-aws) [reply qa/rolling-upgrade/build.gradle:209]: If k-NN is not excluded, then there is no need to exclude batch ingestion test.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1291]: Agree that we can only verify a single document. For `getFirstInnerHit`, we can query the index as the following. This saves us querying the document with the document id.
```
Map<String, Object> searchResults = search(indexName, query, 10);
assertNotNull(searchResults);
Map<String, Object> document = getFirstInnerHit(searchResults);
```

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:694]: See my comment below so that we do not need to query a document with `doc id`.

File: src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java:30]: Please read my comment above

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1298]: You can verify `instanceof Map` from batch ingestion IT. Also for sparse encoding, a `Map` type is not enough. We need to verify type `Map<String, Double>`.

File: qa/rolling-upgrade/build.gradle
Comment (chishui) [reply qa/rolling-upgrade/build.gradle:209]: They are independent, we'd better add based on our own need.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1291]: I get your point, that should also work. The reason why I feel I need the document id is that I create documents with ids in order e.g. "0", "1", "2", "3", "4" etc. I verify the last document through the last id so that I can verify not a random one but the specific one and it should be there and not be cut off.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java]: Refactored some code to reuse the main function

Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java]: Great! Thank you.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1298]: why we need a supplier here?

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java]: Is it a hard requirement to verify a specific document? From my understanding, verifying any document from the index makes sense.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:1298]: Making sense

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java]: For example, some other test unexpectedly reuses your index and stores 10 documents but your tests fail to store documents because of some bugs, but your verification can pass because you still get 10 documents through query. By querying doc of specific id, although similar problem mentioned above is not eliminated, but you reduce the chance to pretty low.

Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java]: Making sense. I will resolve the conversation.

**Review by yuye-aws — APPROVED** [review]

**Comment by yuye-aws** [comment]
Hi @vibrantvarun and @zhichao-aws ! Can you help review this PR? Since I am not the maintainer in this repo, we still need two more approvals.

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (zhichao-aws) [inline src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:717]: Since we have bind the pipeline for index in `createIndexWithConfiguration`, do we still need `pipeline=%s` here?

Comment (chishui) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:717]: It's the same pipeline and just to ensure the pipeline is used, no other impact.

**Review by zhichao-aws — APPROVED** [review]

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (zhichao-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:717]: OK, although I think it's neccessary, but this is not a blocking issue for me. Approved

File: qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/BatchIngestionIT.java
Comment (vibrantvarun) [inline qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/BatchIngestionIT.java:26]: Can this variable be declared at line 22?

File: qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/BatchIngestionIT.java
Comment (vibrantvarun) [inline qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/BatchIngestionIT.java:38]: nit: validateDocInfoAndCount

File: src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java
Comment (vibrantvarun) [inline src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java:695]: nit: bulkAddDocuments

File: src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java
Comment (vibrantvarun) [inline src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java:30]: Please address @yuye-aws comments. 
nit: prepareDataForBulkIngestion

File: qa/rolling-upgrade/build.gradle
Comment (junqiu-lei) [reply qa/rolling-upgrade/build.gradle:209]: Radial search is only available since OpenSearch version 2.14.

File: src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java
Comment (yuye-aws) [reply src/testFixtures/java/org/opensearch/neuralsearch/util/BatchIngestionUtils.java:30]: I am OK with this function.

**Pull request merged**
